gazebo_ros_control/src/default_robot_hw_sim.cpp

```diff
@@ -0,0 +1,654 @@
+                        &joint_effort_limits_[j], &joint_vel_limits_[j]);
+    if (joint_control_methods_[j] != EFFORT)
+    {
+      // Initialize the PID controller. If no PID gain values are found, use joint->SetAngle() or
+      // joint->SetParam("vel") to control the joint.
+#if 0
+      const ros::NodeHandle nh(robot_namespace + "/gazebo_ros_control/pid_gains/" +
+                               joint_names_[j]);
+#endif
+#if 0
+      if (pid_controllers_[j].init(nh))
+      {
+        switch (joint_control_methods_[j])
+        {
+          case POSITION:
+            joint_control_methods_[j] = POSITION_PID;
+            break;
+          case VELOCITY:
+            joint_control_methods_[j] = VELOCITY_PID;
+            break;
+        }
+      }
+      else
+#endif
+      {
+        // joint->SetParam("fmax") must be called if joint->SetAngle() or joint->SetParam("vel") are
+        // going to be called. joint->SetParam("fmax") must *not* be called if joint->SetForce() is
+        // going to be called.
+#if GAZEBO_MAJOR_VERSION > 2
+        joint->SetParam("fmax", 0, joint_effort_limits_[j]);
+#else
+        joint->SetMaxForce(0, joint_effort_limits_[j]);
+#endif
+      }
+
+    }
+    joint_states_[j] = hardware_interface::JointStateHandle(
+      joint_names_[j], &joint_position_[j], &joint_velocity_[j], &joint_effort_[j]);
+    if (register_joint_state_handle(&joint_states_[j]) != hardware_interface::HW_RET_OK) {
+      RCLCPP_WARN_ONCE(LOGGER,"cant register jointstatehandle");
+    }
+
+    joint_cmds_[j] = hardware_interface::JointCommandHandle(
+      joint_names_[j], &joint_position_command_[j]);
+    /*if (register_joint_command_handle(&joint_cmds_[j]) != hardware_interface::HW_RET_OK) {
+      RCLCPP_WARN_ONCE(LOGGER,"cant register jointcommandhandle");
+    }*/
+
+
+    joint_opmodehandles_[j] = hardware_interface::OperationModeHandle(
+      joint_names_[j], &joint_opmodes_[j]);
+    if (register_operation_mode_handle(&joint_opmodehandles_[j]) != hardware_interface::HW_RET_OK) {
+      RCLCPP_WARN_ONCE(LOGGER,"cant register jointopmodehandle");
+    }
+
+    joint_limits_interface::JointLimits limits; //hack, refactor registerjointhandle
+    limits.has_position_limits = true;
+    limits.min_position = joint_lower_limits_[j];
+    limits.max_position = joint_upper_limits_[j];
+    limits.max_velocity = joint_vel_limits_[j];
+    limits.has_velocity_limits = true;
+    limits.max_effort = joint_effort_limits_[j];
+    limits.has_effort_limits = true;
+
+    joint_pos_limit_handles_[j] = joint_limits_interface::PositionJointSaturationHandle(
+      joint_states_[j], joint_cmds_[j], limits);
+
+    joint_eff_cmdhandle_[j] = hardware_interface::JointCommandHandle(
+      joint_names_[j], &joint_effort_command_[j]);
+    //should be register_joint_effort_command_handle, but there's only 1 buffer for now
+    if (register_joint_command_handle(&joint_eff_cmdhandle_[j]) != hardware_interface::HW_RET_OK) {
```

Comment:
The joint commands are registered here. The controllers (joint_trajectory_controller in this case) then use the registered commands. Porters should account for how these values are fed into the controllers.